geomesa-spark/geomesa-spark-jts/src/main/scala/org/apache/spark/sql/jts/JTSTypes.scala

```diff
@@ -44,26 +52,389 @@ object JTSTypes {
+        var j = 0
+        while (j < c.length) {
+          c(j) = new Coordinate(xx.getDouble(j), yy.getDouble(j))
+          j += 1
+        }
+        lines(i) = gf.createLineString(c)
+        i += 1
+      }
+      gf.createMultiLineString(lines)
+    }
+  }
+
+  // case ObjectType.MULTILINESTRING =>
+  //   group.id(GeometryBytes.TwkbMultiLineString)
+  //     .requiredList().element(PrimitiveTypeName.DOUBLE, Repetition.REPEATED).named(GeometryColumnX)
+  //     .requiredList().element(PrimitiveTypeName.DOUBLE, Repetition.REPEATED).named(GeometryColumnY)
+
+
+  //  mline:        OPTIONAL F:2
+  //  .x:           OPTIONAL F:1
+  //  ..list:       REPEATED F:1
+  //  ...element:   OPTIONAL DOUBLE R:1 D:4
+  //  .y:           OPTIONAL F:1
+  //  ..list:       REPEATED F:1
+  //  ...element:   OPTIONAL DOUBLE R:1 D:4
+}
+
+case object MultiLineStringUDT extends MultiLineStringUDT
+
+class PolygonUDT extends AbstractGeometryUDT[Polygon]("polygon") {
+
+  override def sqlType: DataType = StructType(Seq(
+    StructField("x", CoordArray(CoordArray(DataTypes.DoubleType)), nullable = false),
+    StructField("y", CoordArray(CoordArray(DataTypes.DoubleType)), nullable = false)
+  ))
+
+  override def serialize(obj: Polygon): InternalRow = {
+    if (obj == null) { null } else {
+      val x = new GenericArrayData(Array.ofDim[Any](obj.getNumInteriorRing + 1))
+      val y = new GenericArrayData(Array.ofDim[Any](obj.getNumInteriorRing + 1))
+      var i = 0
+      while (i < obj.getNumInteriorRing + 1) {
+        val line = if (i == 0) { obj.getExteriorRing} else { obj.getInteriorRingN(i - 1) }
+        val xx = new GenericArrayData(Array.ofDim[Any](line.getNumPoints))
+        val yy = new GenericArrayData(Array.ofDim[Any](line.getNumPoints))
+        var j = 0
+        while (j < xx.numElements()) {
+          val c = line.getCoordinateN(j)
+          xx(j) = c.getX
+          yy(j) = c.getY
+          j += 1
+        }
+        x(i) = xx
+        y(i) = yy
+        i += 1
+      }
+      new GenericInternalRow(Array[Any](x, y))
+    }
+  }
+
+  override def deserialize(datum: Any): Polygon = {
+    if (datum == null) { null } else {
+      val row = ir(datum)
+      val x = row.getArray(0)
+      val y = row.getArray(1)
+      var shell: LinearRing = null
+      val holes = Array.ofDim[LinearRing](x.numElements() - 1)
+      var i = 0
+      while (i < x.numElements()) {
+        val xx = x.getArray(i)
+        val yy = y.getArray(i)
+        val c = Array.ofDim[Coordinate](xx.numElements())
+        var j = 0
+        while (j < c.length) {
+          c(j) = new Coordinate(xx.getDouble(j), yy.getDouble(j))
+          j += 1
+        }
+        val line = gf.createLinearRing(c)
+        if (i == 0) {
+          shell = line
+        } else {
+          holes(i - 1) = line
+        }
+        i += 1
+      }
+      gf.createPolygon(shell, holes)
+    }
+  }
+
+  // case ObjectType.POLYGON =>
+  //   group.id(GeometryBytes.TwkbPolygon)
```

Comment:
@elahrvivaz what's with the Twkb bit of these variable names?

Reply:
good question... those came from our SimpleFeatureParquetSchema. In checking the code, they correspond to the WKB constants for the same types. We probably don't need to have both pairs of constants, but meh.